Warnings in building wget_agent on GCC8

<!-- Before filling this issue, please read the wiki (https://github.com/fossology/fossology/wiki)
and search if the bug do not already exists in the issues (https://github.com/fossology/fossology/issues). -->

### Logs

```lisp
wget_agent.c: In function ‘GetURL’:
wget_agent.c:319:37: warning: ‘%s’ directive writing up to 2047 bytes into a region of size 2041 [-Wformat-overflow=]
   sprintf(DeleteTempDirCmd, "rm -rf %s", TempFileDirectory);
                                     ^~   ~~~~~~~~~~~~~~~~~
wget_agent.c:319:3: note: ‘sprintf’ output between 8 and 2055 bytes into a destination of size 2048
   sprintf(DeleteTempDirCmd, "rm -rf %s", TempFileDirectory);
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
wget_agent.c:378:29: warning: ‘%s’ directive output may be truncated writing up to 2047 bytes into a region of size 2046 [-Wformat-truncation=]
     snprintf(CMD,MAXCMD-1," %s /usr/bin/wget -q %s -P '%s' '%s' %s %s 2>&1",
                             ^~
         proxy, WgetArgs,TempFileDirectory,TaintedURL,GlobalParam, no_proxy);
         ~~~~~                
wget_agent.c:378:5: note: ‘snprintf’ output between 36 and 10330 bytes into a destination of size 2047
     snprintf(CMD,MAXCMD-1," %s /usr/bin/wget -q %s -P '%s' '%s' %s %s 2>&1",
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
         proxy, WgetArgs,TempFileDirectory,TaintedURL,GlobalParam, no_proxy);
         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
wget_agent.c:383:29: warning: ‘%s’ directive output may be truncated writing up to 2047 bytes into a region of size 2046 [-Wformat-truncation=]
     snprintf(CMD,MAXCMD-1," %s /usr/bin/wget -q %s -P '%s' '%s' %s %s 2>&1",
                             ^~
         proxy, WgetArgs, TempFileDir, TaintedURL, GlobalParam, no_proxy);
         ~~~~~                
wget_agent.c:383:5: note: ‘snprintf’ output 36 or more bytes (assuming 8283) into a destination of size 2047
     snprintf(CMD,MAXCMD-1," %s /usr/bin/wget -q %s -P '%s' '%s' %s %s 2>&1",
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
         proxy, WgetArgs, TempFileDir, TaintedURL, GlobalParam, no_proxy);
         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
wget_agent.c:388:29: warning: ‘%s’ directive output may be truncated writing up to 2047 bytes into a region of size 2046 [-Wformat-truncation=]
     snprintf(CMD,MAXCMD-1," %s /usr/bin/wget -q %s '%s' %s %s 2>&1",
                             ^~
         proxy, WgetArgs,TaintedURL, GlobalParam, no_proxy);
         ~~~~~                
wget_agent.c:388:5: note: ‘snprintf’ output between 30 and 8277 bytes into a destination of size 2047
     snprintf(CMD,MAXCMD-1," %s /usr/bin/wget -q %s '%s' %s %s 2>&1",
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
         proxy, WgetArgs,TaintedURL, GlobalParam, no_proxy);
         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
wget_agent.c:424:41: warning: ‘/’ directive output may be truncated writing 1 byte into a region of size between 0 and 2047 [-Wformat-truncation=]
     snprintf(TempFilePath, MAXCMD-1, "%s/%s", TempFileDirectory, TaintedURL + Position);
                                         ^
wget_agent.c:424:5: note: ‘snprintf’ output 2 or more bytes (assuming 2049) into a destination of size 2047
     snprintf(TempFilePath, MAXCMD-1, "%s/%s", TempFileDirectory, TaintedURL + Position);
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
wget_agent.c:454:37: warning: ‘%s’ directive output may be truncated writing up to 2047 bytes into a region of size 2041 [-Wformat-truncation=]
       snprintf(CMD,MAXCMD-1, "find '%s' -type f -exec mv {} %s \\; > /dev/null 2>&1", TempFileDirectory, TempFile);
                                     ^~                                                ~~~~~~~~~~~~~~~~~
wget_agent.c:454:7: note: ‘snprintf’ output 49 or more bytes (assuming 2096) into a destination of size 2047
       snprintf(CMD,MAXCMD-1, "find '%s' -type f -exec mv {} %s \\; > /dev/null 2>&1", TempFileDirectory, TempFile);
       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
wget_agent.c:439:37: warning: ‘%s’ directive output may be truncated writing up to 2047 bytes into a region of size 2043 [-Wformat-truncation=]
         snprintf(CMD,MAXCMD-1, "mv '%s' '%s' 2>&1", TempFilePath, TempFile);
                                     ^~              ~~~~~~~~~~~~
wget_agent.c:439:9: note: ‘snprintf’ output 14 or more bytes (assuming 2061) into a destination of size 2047
         snprintf(CMD,MAXCMD-1, "mv '%s' '%s' 2>&1", TempFilePath, TempFile);
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
wget_agent.c:431:39: warning: ‘%s’ directive output may be truncated writing up to 2047 bytes into a region of size 2041 [-Wformat-truncation=]
         snprintf(CMD,MAXCMD-1, "find '%s' -mindepth 1 -type d -empty -exec rmdir {} \\; > /dev/null 2>&1", TempFilePath);
                                       ^~                                                                   ~~~~~~~~~~~~
wget_agent.c:431:9: note: ‘snprintf’ output between 70 and 2117 bytes into a destination of size 2047
         snprintf(CMD,MAXCMD-1, "find '%s' -mindepth 1 -type d -empty -exec rmdir {} \\; > /dev/null 2>&1", TempFilePath);
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
wget_agent.c:435:51: warning: ‘%s’ directive output may be truncated writing up to 2047 bytes into a region of size 2031 [-Wformat-truncation=]
         snprintf(CMD,MAXCMD-1, "tar -cf  '%s' -C '%s' ./ 1>/dev/null", TempFile, TempFilePath);
                                                   ^~                             ~~~~~~~~~~~~
wget_agent.c:435:9: note: ‘snprintf’ output 33 or more bytes (assuming 2080) into a destination of size 2047
         snprintf(CMD,MAXCMD-1, "tar -cf  '%s' -C '%s' ./ 1>/dev/null", TempFile, TempFilePath);
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
wget_agent.c: In function ‘SetEnv’:
wget_agent.c:638:3: warning: ‘strncpy’ specified bound 2048 equals destination size [-Wstringop-truncation]
   strncpy(GlobalParam, S, sizeof(GlobalParam)); // get the parameters, kind of " -A rpm -R fosso -l 1* "
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
wget_agent.c: In function ‘Archivefs’:
wget_agent.c:714:33: warning: ‘%s’ directive output may be truncated writing up to 2047 bytes into a region of size 2043 [-Wformat-truncation=]
     snprintf(CMD,MAXCMD-1, "tar %s -cf  '%s' -C '%s' ./ 1>/dev/null", GlobalParam, TempFile, Path);
                                 ^~                                    ~~~~~~~~~~~
wget_agent.c:714:5: note: ‘snprintf’ output 34 or more bytes (assuming 2081) into a destination of size 2047
     snprintf(CMD,MAXCMD-1, "tar %s -cf  '%s' -C '%s' ./ 1>/dev/null", GlobalParam, TempFile, Path);
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
wget_agent.c: In function ‘replace_url_with_auth’:
wget_agent.c:880:44: warning: ‘%s’ directive output may be truncated writing up to 2047 bytes into a region of size between 2037 and 2046 [-Wformat-truncation=]
     snprintf(GlobalURL, FILEPATH, "%s%s:%s@%s", http, username, password, URI);
                                            ^~                             ~~~
wget_agent.c:880:5: note: ‘snprintf’ output 3 or more bytes (assuming 2059) into a destination of size 2048
     snprintf(GlobalURL, FILEPATH, "%s%s:%s@%s", http, username, password, URI);
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
wget_agent.c:868:5: warning: ‘strncpy’ specified bound depends on the length of the source argument [-Wstringop-overflow=]
     strncpy(http, GlobalURL, strlen(GlobalURL) - strlen(URI));
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
wget_agent.c:868:30: note: length computed here
     strncpy(http, GlobalURL, strlen(GlobalURL) - strlen(URI));
                              ^~~~~~~~~~~~~~~~~
wget_agent.c: In function ‘GetVersionControlCommand’:
wget_agent.c:950:33: warning: ‘.dir’ directive writing 4 bytes into a region of size between 1 and 2048 [-Wformat-overflow=]
   sprintf(TempFileDirectory, "%s.dir", GlobalTempFile);
                                 ^~~~
wget_agent.c:950:3: note: ‘sprintf’ output between 5 and 2052 bytes into a destination of size 2048
   sprintf(TempFileDirectory, "%s.dir", GlobalTempFile);
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
wget_agent.c:967:72: warning: ‘%s’ directive writing up to 2047 bytes into a region of size 2003 [-Wformat-overflow=]
       sprintf(command, "git config --global http.proxy %s && git clone %s %s %s  && rm -rf %s/.git", GlobalProxy[0], GlobalURL, GlobalParam, TempFileDirectory, TempFileDirectory);
                                                                        ^~                                            ~~~~~~~~~
wget_agent.c:967:7: note: ‘sprintf’ output 65 or more bytes (assuming 8253) into a destination of size 2048
       sprintf(command, "git config --global http.proxy %s && git clone %s %s %s  && rm -rf %s/.git", GlobalProxy[0], GlobalURL, GlobalParam, TempFileDirectory, TempFileDirectory);
       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
wget_agent.c:969:35: warning: ‘%s’ directive writing up to 2047 bytes into a region of size 2038 [-Wformat-overflow=]
       sprintf(command, "git clone %s %s %s >/dev/null 2>&1 && rm -rf %s/.git", GlobalURL, GlobalParam, TempFileDirectory, TempFileDirectory);
                                   ^~                                           ~~~~~~~~~
wget_agent.c:969:7: note: ‘sprintf’ output between 45 and 8233 bytes into a destination of size 2048
       sprintf(command, "git clone %s %s %s >/dev/null 2>&1 && rm -rf %s/.git", GlobalURL, GlobalParam, TempFileDirectory, TempFileDirectory);
       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
wget_agent.c:959:136: warning: ‘%s’ directive writing up to 2047 bytes into a region of size 1941 [-Wformat-overflow=]
       sprintf(command, "svn --config-option servers:global:http-proxy-host=%s --config-option servers:global:http-proxy-port=%s export %s %s %s --no-auth-cache >/dev/null 2>&1", GlobalProxy[4], GlobalProxy[5], GlobalURL, GlobalParam, TempFileDirectory);
                                                                                                                                        ^~                                                                         ~~~~~~~~~
wget_agent.c:959:7: note: ‘sprintf’ output 142 or more bytes (assuming 6283) into a destination of size 2048
       sprintf(command, "svn --config-option servers:global:http-proxy-host=%s --config-option servers:global:http-proxy-port=%s export %s %s %s --no-auth-cache >/dev/null 2>&1", GlobalProxy[4], GlobalProxy[5], GlobalURL, GlobalParam, TempFileDirectory);
       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
wget_agent.c:961:36: warning: ‘%s’ directive writing up to 2047 bytes into a region of size 2037 [-Wformat-overflow=]
       sprintf(command, "svn export %s %s %s --no-auth-cache >/dev/null 2>&1", GlobalURL, GlobalParam, TempFileDirectory);
                                    ^~                                         ~~~~~~~~~
wget_agent.c:961:7: note: ‘sprintf’ output between 46 and 6187 bytes into a destination of size 2048
       sprintf(command, "svn export %s %s %s --no-auth-cache >/dev/null 2>&1", GlobalURL, GlobalParam, TempFileDirectory);
       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
wget_agent.c: In function ‘GetVersionControl’:
wget_agent.c:508:33: warning: ‘.dir’ directive writing 4 bytes into a region of size between 1 and 2048 [-Wformat-overflow=]
   sprintf(TempFileDirectory, "%s.dir", GlobalTempFile);
                                 ^~~~
wget_agent.c:508:3: note: ‘sprintf’ output between 5 and 2052 bytes into a destination of size 2048
   sprintf(TempFileDirectory, "%s.dir", GlobalTempFile);
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
wget_agent.c:509:37: warning: ‘%s’ directive writing up to 2047 bytes into a region of size 2041 [-Wformat-overflow=]
   sprintf(DeleteTempDirCmd, "rm -rf %s", TempFileDirectory);
                                     ^~   ~~~~~~~~~~~~~~~~~
wget_agent.c:509:3: note: ‘sprintf’ output between 8 and 2055 bytes into a destination of size 2048
   sprintf(DeleteTempDirCmd, "rm -rf %s", TempFileDirectory);
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
wget_agent.c:508:33: warning: ‘.dir’ directive writing 4 bytes into a region of size between 1 and 2048 [-Wformat-overflow=]
   sprintf(TempFileDirectory, "%s.dir", GlobalTempFile);
                                 ^~~~
wget_agent.c:508:3: note: ‘sprintf’ output between 5 and 2052 bytes into a destination of size 2048
   sprintf(TempFileDirectory, "%s.dir", GlobalTempFile);
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
wget_agent.c:509:37: warning: ‘%s’ directive writing up to 2047 bytes into a region of size 2041 [-Wformat-overflow=]
   sprintf(DeleteTempDirCmd, "rm -rf %s", TempFileDirectory);
                                     ^~   ~~~~~~~~~~~~~~~~~
wget_agent.c:509:3: note: ‘sprintf’ output between 8 and 2055 bytes into a destination of size 2048
   sprintf(DeleteTempDirCmd, "rm -rf %s", TempFileDirectory);
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
wget_agent.c:535:41: warning: ‘%s’ directive output may be truncated writing up to 2047 bytes into a region of size 2037 [-Wformat-truncation=]
   snprintf(command,MAXCMD-1, "tar -cf  '%s' -C '%s' ./ 1>/dev/null", GlobalTempFile, TempFileDirectory);
                                         ^~                           ~~~~~~~~~~~~~~
wget_agent.c:535:3: note: ‘snprintf’ output between 33 and 4127 bytes into a destination of size 2047
   snprintf(command,MAXCMD-1, "tar -cf  '%s' -C '%s' ./ 1>/dev/null", GlobalTempFile, TempFileDirectory);
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
main.c: In function ‘main’:
main.c:211:5: warning: ‘strncpy’ specified bound 2048 equals destination size [-Wstringop-truncation]
     strncpy(GlobalURL,argv[arg],sizeof(GlobalURL));
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

--- a/src/wget_agent/agent/main.c
+++ b/src/wget_agent/agent/main.c
@@ -115,16 +115,16 @@ int main  (int argc, char *argv[])
   int user_pk;
   char *agent_desc = "Network downloader.  Uses wget(1).";
 
-  memset(GlobalTempFile,'\0',MAXCMD);
-  memset(GlobalURL,'\0',MAXCMD);
-  memset(GlobalParam,'\0',MAXCMD);
-  memset(GlobalType,'\0',MAXCMD);
+  memset(GlobalTempFile,'\0',STRMAX);
+  memset(GlobalURL,'\0',URLMAX);
+  memset(GlobalParam,'\0',STRMAX);
+  memset(GlobalType,'\0',STRMAX);
   GlobalUploadKey = -1;
   int upload_pk = 0;           // the upload primary key
   //int Agent_pk;
   char *COMMIT_HASH;
   char *VERSION;
-  char agent_rev[MAXCMD];
+  char agent_rev[STRMAX];
 
   /* open the connection to the scheduler and configuration */
   fo_scheduler_connect(&argc, argv, &pgConn);
@@ -156,16 +156,16 @@ int main  (int argc, char *argv[])
           strcpy(GlobalTempFile,"wget.default_download");
         break;
       case 'A':
-        strncat(GlobalParam, " -A ", MAXCMD - strlen(GlobalParam) -1);
-        strncat(GlobalParam, optarg, MAXCMD - strlen(GlobalParam) -1);
+        strncat(GlobalParam, " -A ", STRMAX - strlen(GlobalParam) -1);
+        strncat(GlobalParam, optarg, STRMAX - strlen(GlobalParam) -1);
         break;
       case 'R':
-        strncat(GlobalParam, " -R ", MAXCMD - strlen(GlobalParam) -1);
-        strncat(GlobalParam, optarg, MAXCMD - strlen(GlobalParam) -1);
+        strncat(GlobalParam, " -R ", STRMAX - strlen(GlobalParam) -1);
+        strncat(GlobalParam, optarg, STRMAX - strlen(GlobalParam) -1);
         break;
       case 'l':
-        strncat(GlobalParam, " -l ", MAXCMD - strlen(GlobalParam) -1);
-        strncat(GlobalParam, optarg, MAXCMD - strlen(GlobalParam) -1);
+        strncat(GlobalParam, " -l ", STRMAX - strlen(GlobalParam) -1);
+        strncat(GlobalParam, optarg, STRMAX - strlen(GlobalParam) -1);
         break;
       case 'c': break; /* handled by fo_scheduler_connect() */
       case 'C':
@@ -229,7 +229,7 @@ int main  (int argc, char *argv[])
       LOG_VERBOSE0("It's a file -- GlobalUploadKey = %ld",GlobalUploadKey);
       if (GlobalUploadKey != -1)
       {
-        memcpy(GlobalTempFile,GlobalURL,MAXCMD);
+        memcpy(GlobalTempFile,GlobalURL,STRMAX);
         DBLoadGold();
       }
     }
@@ -256,9 +256,9 @@ int main  (int argc, char *argv[])
           continue;
         }
 
-        char TempDir[MAXCMD];
-        memset(TempDir,'\0',MAXCMD);
-        snprintf(TempDir, MAXCMD-1, "%s/wget", TempFileDir); // /var/local/lib/fossology/agents/wget
+        char TempDir[STRMAX];
+        memset(TempDir,'\0',STRMAX);
+        snprintf(TempDir, STRMAX-1, "%s/wget", TempFileDir); // /var/local/lib/fossology/agents/wget
         struct stat Status = {};
 
         if (GlobalType[0])

--- a/src/wget_agent/agent/wget_agent.c
+++ b/src/wget_agent/agent/wget_agent.c
@@ -372,15 +372,20 @@ int GetURL(char *TempFile, char *URL, char *TempFileDir)
   char *cmd;
   char TaintedURL[MAXCMD];
   char TempFileDirectory[MAXCMD];
-  char DeleteTempDirCmd[MAXCMD];
+  char *delete_tmpdir_cmd;
   int rc;
+  int res;
 
   memset(TempFileDirectory,'\0',MAXCMD);
-  memset(DeleteTempDirCmd,'\0',MAXCMD);
 
   /* save each upload files in /srv/fossology/repository/localhost/wget/wget.xxx.dir/ */
   sprintf(TempFileDirectory, "%s.dir", TempFile);
-  sprintf(DeleteTempDirCmd, "rm -rf %s", TempFileDirectory);
+  res = asprintf(&delete_tmpdir_cmd, "rm -rf %s", TempFileDirectory);
+  if (res == -1)
+  {
+    ASPRINTF_MEM_ERROR_LOG;
+    SafeExit(ASPRINTF_MEM_ERROR);
+  }
 #if 1
   char WgetArgs[]="--no-check-certificate --progress=dot -rc -np -e robots=off";
 #else
@@ -435,7 +440,6 @@ int GetURL(char *TempFile, char *URL, char *TempFileDir)
   }
 
   char *dest;
-  int res;
 
   dest = PrepareWgetDest(TempFile, TempFileDir, TempFileDirectory);
 
@@ -452,6 +456,7 @@ int GetURL(char *TempFile, char *URL, char *TempFileDir)
   if (res == -1)
   {
     ASPRINTF_MEM_ERROR_LOG;
+    free(delete_tmpdir_cmd);
     SafeExit(ASPRINTF_MEM_ERROR);
   }
 
@@ -468,8 +473,9 @@ int GetURL(char *TempFile, char *URL, char *TempFileDir)
   {
     LOG_FATAL("upload %ld Download failed; Return code %d from: %s",GlobalUploadKey,WEXITSTATUS(rc),cmd);
     unlink(GlobalTempFile);
-    rc_system = system(DeleteTempDirCmd);
-    if (!WIFEXITED(rc_system)) systemError(__LINE__, rc_system, DeleteTempDirCmd)
+    rc_system = system(delete_tmpdir_cmd);
+    if (!WIFEXITED(rc_system)) systemError(__LINE__, rc_system, delete_tmpdir_cmd)
+    free(delete_tmpdir_cmd);
     SafeExit(12);
   }
 
@@ -484,8 +490,9 @@ int GetURL(char *TempFile, char *URL, char *TempFileDir)
     {
       LOG_FATAL("path %s is not http://, https://, or ftp://", TaintedURL);
       unlink(GlobalTempFile);
-      rc_system = system(DeleteTempDirCmd);
-      if (!WIFEXITED(rc_system)) systemError(__LINE__, rc_system, DeleteTempDirCmd)
+      rc_system = system(delete_tmpdir_cmd);
+      if (!WIFEXITED(rc_system)) systemError(__LINE__, rc_system, delete_tmpdir_cmd)
+      free(delete_tmpdir_cmd);
       SafeExit(26);
     }
     snprintf(TempFilePath, MAXCMD-1, "%s/%s", TempFileDirectory, TaintedURL + Position);
@@ -498,6 +505,7 @@ int GetURL(char *TempFile, char *URL, char *TempFileDir)
         if (res == -1)
         {
           ASPRINTF_MEM_ERROR_LOG;
+          free(delete_tmpdir_cmd);
           SafeExit(ASPRINTF_MEM_ERROR);
         }
         rc_system = system(cmd); // delete all empty directories downloaded
@@ -508,6 +516,7 @@ int GetURL(char *TempFile, char *URL, char *TempFileDir)
         if (res == -1)
         {
           ASPRINTF_MEM_ERROR_LOG;
+          free(delete_tmpdir_cmd);
           SafeExit(ASPRINTF_MEM_ERROR);
         }
       }
@@ -517,6 +526,7 @@ int GetURL(char *TempFile, char *URL, char *TempFileDir)
         if (res == -1)
         {
           ASPRINTF_MEM_ERROR_LOG;
+          free(delete_tmpdir_cmd);
           SafeExit(ASPRINTF_MEM_ERROR);
         }
       }
@@ -527,8 +537,9 @@ int GetURL(char *TempFile, char *URL, char *TempFileDir)
         systemError(__LINE__, rc_system, cmd)
         free(cmd);
         unlink(GlobalTempFile);
-        rc_system = system(DeleteTempDirCmd);
-        if (!WIFEXITED(rc_system)) systemError(__LINE__, rc_system, DeleteTempDirCmd)
+        rc_system = system(delete_tmpdir_cmd);
+        if (!WIFEXITED(rc_system)) systemError(__LINE__, rc_system, delete_tmpdir_cmd)
+        free(delete_tmpdir_cmd);
         SafeExit(24); // failed to store the temperary directory(one file) as one temperary file
       }
 
@@ -539,6 +550,7 @@ int GetURL(char *TempFile, char *URL, char *TempFileDir)
       if (res == -1)
       {
         ASPRINTF_MEM_ERROR_LOG;
+        free(delete_tmpdir_cmd);
         SafeExit(ASPRINTF_MEM_ERROR);
       }
       rc_system = system(cmd);
@@ -547,8 +559,9 @@ int GetURL(char *TempFile, char *URL, char *TempFileDir)
         systemError(__LINE__, rc_system, cmd)
         free(cmd);
         unlink(GlobalTempFile);
-        rc_system = system(DeleteTempDirCmd);
-        if (!WIFEXITED(rc_system)) systemError(__LINE__, rc_system, DeleteTempDirCmd)
+        rc_system = system(delete_tmpdir_cmd);
+        if (!WIFEXITED(rc_system)) systemError(__LINE__, rc_system, delete_tmpdir_cmd)
+        free(delete_tmpdir_cmd);
         SafeExit(24); // failed to store the temperary directory(one file) as one temperary file
       }
 
@@ -560,17 +573,21 @@ int GetURL(char *TempFile, char *URL, char *TempFileDir)
     LOG_FATAL("upload %ld File %s not created from URL: %s, CMD: %s",GlobalUploadKey,TempFile,URL, cmd);
     free(cmd);
     unlink(GlobalTempFile);
-    rc_system = system(DeleteTempDirCmd);
-    if (!WIFEXITED(rc_system)) systemError(__LINE__, rc_system, DeleteTempDirCmd)
+    rc_system = system(delete_tmpdir_cmd);
+    if (!WIFEXITED(rc_system)) systemError(__LINE__, rc_system, delete_tmpdir_cmd)
+    free(delete_tmpdir_cmd);
     SafeExit(15);
   }
   
   free(cmd);
 
   /* remove the temp dir /srv/fossology/repository/localhost/wget/wget.xxx.dir/ for this upload */
-  rc_system = system(DeleteTempDirCmd);
-  if (!WIFEXITED(rc_system)) systemError(__LINE__, rc_system, DeleteTempDirCmd)
+  rc_system = system(delete_tmpdir_cmd);
+  if (!WIFEXITED(rc_system)) systemError(__LINE__, rc_system, delete_tmpdir_cmd)
   LOG_VERBOSE0("upload %ld Downloaded %s to %s",GlobalUploadKey,URL,TempFile);
+
+  free(delete_tmpdir_cmd);
+
   return(0);
 } /* GetURL() */
 

--- a/src/wget_agent/agent/wget_agent.c
+++ b/src/wget_agent/agent/wget_agent.c
@@ -117,14 +117,14 @@ void DBLoadGold()
     SafeExit(ASPRINTF_MEM_ERROR);
   }
 
-  FILE* file = popen(cmd, "r");
-
   if (!Fin)
   {
     LOG_FATAL("upload %ld Unable to open temp file %s from %s",
         GlobalUploadKey,GlobalTempFile,GlobalURL);
     SafeExit(1);
   }
+  
+  FILE* file = popen(cmd, "r");
 
   if (!file)
   {

--- a/src/wget_agent/agent/wget_agent.c
+++ b/src/wget_agent/agent/wget_agent.c
@@ -29,16 +29,16 @@
 
 #include "wget_agent.h"
 
-char SQL[MAXCMD];
+char SQL[STRMAX];
 
 PGconn *pgConn = NULL;        ///< For the DB
 long GlobalUploadKey=-1;      ///< Input for this system
-char GlobalTempFile[MAXCMD];  ///< Temp file to be used
-char GlobalURL[MAXCMD];       ///< URL to download
-char GlobalType[MAXCMD];      ///< Type of download (FILE/version control)
-char GlobalParam[MAXCMD];     ///< Additional parameters
+char GlobalTempFile[STRMAX];  ///< Temp file to be used
+char GlobalURL[URLMAX];       ///< URL to download
+char GlobalType[STRMAX];      ///< Type of download (FILE/version control)
+char GlobalParam[STRMAX];     ///< Additional parameters
 char *GlobalProxy[6];         ///< Proxy from fossology.conf
-char GlobalHttpProxy[MAXCMD]; ///< HTTP proxy command to use
+char GlobalHttpProxy[STRMAX]; ///< HTTP proxy command to use
 int GlobalImportGold=1;       ///< Set to 0 to not store file in gold repository
 gid_t ForceGroup=-1;          ///< Set to group id to be used for download files
 
@@ -95,7 +95,7 @@ void DBLoadGold()
   Cksum *Sum;
   char *Unique=NULL;
   char *SHA1, *MD5, *Len;
-  char SQL[MAXCMD];
+  char SQL[STRMAX];
   long PfileKey;
   char *Path;
   char SHA256[65];
@@ -239,8 +239,8 @@ void DBLoadGold()
   MD5 = Unique+41; /* 40 for sha1 + 1 for '.' */
   Len = Unique+41+33; /* 32 for md5 + 1 for '.' */
   /* Set the pfile */
-  memset(SQL,'\0',MAXCMD);
-  snprintf(SQL,MAXCMD-1,"SELECT pfile_pk FROM pfile WHERE pfile_sha1 = '%.40s' AND pfile_md5 = '%.32s' AND pfile_size = %s;",
+  memset(SQL,'\0',STRMAX);
+  snprintf(SQL,STRMAX-1,"SELECT pfile_pk FROM pfile WHERE pfile_sha1 = '%.40s' AND pfile_md5 = '%.32s' AND pfile_size = %s;",
       SHA1,MD5,Len);
   result =  PQexec(pgConn, SQL); /* SELECT */
   if (fo_checkPQresult(pgConn, result, SQL, __FILE__, __LINE__)) SafeExit(7);
@@ -249,8 +249,8 @@ void DBLoadGold()
   if (PQntuples(result) <=0)
   {
     /* Insert it */
-    memset(SQL,'\0',MAXCMD);
-    snprintf(SQL,MAXCMD-1,"INSERT INTO pfile (pfile_sha1, pfile_md5, pfile_sha256, pfile_size) VALUES ('%.40s','%.32s','%.64s',%s)",
+    memset(SQL,'\0',STRMAX);
+    snprintf(SQL,STRMAX-1,"INSERT INTO pfile (pfile_sha1, pfile_md5, pfile_sha256, pfile_size) VALUES ('%.40s','%.32s','%.64s',%s)",
         SHA1,MD5,SHA256,Len);
     PQclear(result);
     result = PQexec(pgConn, SQL);
@@ -268,14 +268,14 @@ void DBLoadGold()
   result = PQexec(pgConn, "BEGIN");
   if (fo_checkPQcommand(pgConn, result, SQL, __FILE__, __LINE__)) SafeExit(-1);
 
-  memset(SQL,0,MAXCMD);
-  snprintf(SQL,MAXCMD-1,"SELECT * FROM upload WHERE upload_pk=%ld FOR UPDATE;",GlobalUploadKey);
+  memset(SQL,0,STRMAX);
+  snprintf(SQL,STRMAX-1,"SELECT * FROM upload WHERE upload_pk=%ld FOR UPDATE;",GlobalUploadKey);
   PQclear(result);
   result = PQexec(pgConn, SQL);
   if (fo_checkPQresult(pgConn, result, SQL, __FILE__, __LINE__)) SafeExit(-1);
 
-  memset(SQL,0,MAXCMD);
-  snprintf(SQL,MAXCMD-1,"UPDATE upload SET pfile_fk=%ld WHERE upload_pk=%ld",
+  memset(SQL,0,STRMAX);
+  snprintf(SQL,STRMAX-1,"UPDATE upload SET pfile_fk=%ld WHERE upload_pk=%ld",
       PfileKey,GlobalUploadKey);
   LOG_VERBOSE0("SQL=%s\n",SQL);
   PQclear(result);
@@ -370,13 +370,13 @@ char *PrepareWgetDest(char *TempFile, char *TempFileDir, char *TempFileDirectory
 int GetURL(char *TempFile, char *URL, char *TempFileDir)
 {
   char *cmd;
-  char TaintedURL[MAXCMD];
-  char TempFileDirectory[MAXCMD];
+  char TaintedURL[STRMAX];
+  char TempFileDirectory[STRMAX+128];
   char *delete_tmpdir_cmd;
   int rc;
   int res;
 
-  memset(TempFileDirectory,'\0',MAXCMD);
+  memset(TempFileDirectory,'\0',STRMAX+128);
 
   /* save each upload files in /srv/fossology/repository/localhost/wget/wget.xxx.dir/ */
   sprintf(TempFileDirectory, "%s.dir", TempFile);
@@ -393,7 +393,7 @@ int GetURL(char *TempFile, char *URL, char *TempFileDir)
   char WgetArgs[]="--progress=dot -rc -np -e robots=off";
 #endif
 
-  if (!TaintURL(URL,TaintedURL,MAXCMD))
+  if (!TaintURL(URL,TaintedURL,STRMAX))
   {
     LOG_FATAL("Failed to taint the URL '%s'",URL);
     SafeExit(10);
@@ -413,30 +413,30 @@ int GetURL(char *TempFile, char *URL, char *TempFileDir)
 
   struct stat sb;
   int rc_system =0;
-  char no_proxy[MAXCMD] = {0};
-  char proxy[MAXCMD] = {0};
-  char proxy_temp[MAXCMD] = {0};
+  char no_proxy[STRMAX] = {0};
+  char proxy[STRMAX] = {0};
+  char proxy_temp[STRMAX] = {0};
 
   /* http_proxy is optional so don't error if it doesn't exist */
   /** set proxy */
   if (GlobalProxy[0] && GlobalProxy[0][0])
   {
-    snprintf(proxy_temp, MAXCMD-1, "export http_proxy='%s' ;", GlobalProxy[0]);
+    snprintf(proxy_temp, STRMAX-1, "export http_proxy='%s' ;", GlobalProxy[0]);
     strcat(proxy, proxy_temp);
   }
   if (GlobalProxy[1] && GlobalProxy[1][0])
   {
-    snprintf(proxy_temp, MAXCMD-1, "export https_proxy='%s' ;", GlobalProxy[1]);
+    snprintf(proxy_temp, STRMAX-1, "export https_proxy='%s' ;", GlobalProxy[1]);
     strcat(proxy, proxy_temp);
   }
   if (GlobalProxy[2] && GlobalProxy[2][0])
   {
-    snprintf(proxy_temp, MAXCMD-1, "export ftp_proxy='%s' ;", GlobalProxy[2]);
+    snprintf(proxy_temp, STRMAX-1, "export ftp_proxy='%s' ;", GlobalProxy[2]);
     strcat(proxy, proxy_temp);
   }
   if (GlobalProxy[3] && GlobalProxy[3][0])
   {
-    snprintf(no_proxy, MAXCMD-1, "-e no_proxy='%s'", GlobalProxy[3]);
+    snprintf(no_proxy, STRMAX-1, "-e no_proxy='%s'", GlobalProxy[3]);
   }
 
   char *dest;
@@ -636,19 +636,26 @@ int GetVersionControl()
   res = asprintf(&tmp_file_directory, "%s.dir", GlobalTempFile);
   if (res == -1)
   {
+    ASPRINTF_MEM_ERROR_LOG;
     return ASPRINTF_MEM_ERROR;
   }
 
   res = asprintf(&delete_tmpdir_cmd, "rm -rf %s", tmp_file_directory);
   if (res == -1)
   {
+    ASPRINTF_MEM_ERROR_LOG;
     free(tmp_file_directory);
     return ASPRINTF_MEM_ERROR;
   }
 
   command = GetVersionControlCommand(1);
-
+  if (!command)
+  {
+    free(tmp_file_directory);
+    return ASPRINTF_MEM_ERROR;
+  }
   rc = system(command);
+  free(command);
 
   if (resethome) // rollback
     unsetenv("HOME");
@@ -658,6 +665,12 @@ int GetVersionControl()
   if (rc != 0)
   {
     command = GetVersionControlCommand(-1);
+    if (!command)
+    {
+      ASPRINTF_MEM_ERROR_LOG;
+      free(tmp_file_directory);
+      return ASPRINTF_MEM_ERROR;
+    }
     systemError(__LINE__, rc, command)
     /** for user fossy
     \code git: git config --global http.proxy web-proxy.cce.hp.com:8088; git clone http://github.com/schacon/grit.git
@@ -667,12 +680,20 @@ int GetVersionControl()
     /* remove the temp dir /srv/fossology/repository/localhost/wget/wget.xxx.dir/ for this upload */
     rc = system(delete_tmpdir_cmd);
     if (!WIFEXITED(rc)) systemError(__LINE__, rc, delete_tmpdir_cmd)
+    free(command);
     free(tmp_file_directory);
     free(delete_tmpdir_cmd);
     return 1;
   }
 
-  snprintf(command,MAXCMD-1, "tar -cf  '%s' -C '%s' ./ 1>/dev/null", GlobalTempFile, tmp_file_directory);
+  res = asprintf(&command, "tar -cf  '%s' -C '%s' ./ 1>/dev/null", GlobalTempFile, tmp_file_directory);
+  if (res == -1)
+  {
+    ASPRINTF_MEM_ERROR_LOG;
+    free(tmp_file_directory);
+    free(delete_tmpdir_cmd);
+    return ASPRINTF_MEM_ERROR;
+  }
   free(tmp_file_directory);
   rc = system(command);
   if (rc != 0)
@@ -706,8 +727,8 @@ void    SetEnv  (char *S, char *TempFileDir)
   int SLen,GLen; /* lengths for S and global string */
 
   GlobalUploadKey = -1;
-  memset(GlobalTempFile,'\0',MAXCMD);
-  memset(GlobalURL,'\0',MAXCMD);
+  memset(GlobalTempFile,'\0',STRMAX);
+  memset(GlobalURL,'\0',URLMAX);
   if (!S) return;
 
   /* first value is the upload_pk */
@@ -720,7 +741,7 @@ void    SetEnv  (char *S, char *TempFileDir)
   /** This will be removed when the jobqueue is changed. **/
   SLen=0;
   GLen=0;
-  while((GLen < MAXCMD-4) && S[SLen] && !isspace(S[SLen]))
+  while((GLen < STRMAX-4) && S[SLen] && !isspace(S[SLen]))
   {
     if ((S[SLen] == '\'') || isspace(S[SLen]) || !isprint(S[SLen]))
     {
@@ -735,14 +756,14 @@ void    SetEnv  (char *S, char *TempFileDir)
 #endif
   if (TempFileDir)
   {
-    memset(GlobalTempFile,'\0',MAXCMD);
-    snprintf(GlobalTempFile,MAXCMD-1,"%s/wget.%d",TempFileDir,getpid());
+    memset(GlobalTempFile,'\0',STRMAX);
+    snprintf(GlobalTempFile,STRMAX-1,"%s/wget.%d",TempFileDir,getpid());
   }
 
   /* third value is the URL location -- taint any single-quotes */
   SLen=0;
   GLen=0;
-  while((GLen < MAXCMD-4) && S[SLen])
+  while((GLen < STRMAX-4) && S[SLen])
   {
     if ((S[SLen] == '\\') && isprint(S[SLen+1])) // in file path, if include '\ ', that mean this file name include spaces
     {
@@ -767,7 +788,7 @@ void    SetEnv  (char *S, char *TempFileDir)
   char Type[][4] = {"SVN", "Git", "CVS"};
   int i = 0; // type index
 
-  memset(GlobalType,'\0',MAXCMD);
+  memset(GlobalType,'\0',STRMAX);
   strncpy(GlobalType, S, 3);
   if ((0 == strcmp(GlobalType, Type[i++])) || (0 == strcmp(GlobalType, Type[i++])) || (0 == strcmp(GlobalType, Type[i++])))
   {
@@ -775,7 +796,7 @@ void    SetEnv  (char *S, char *TempFileDir)
   }
   else
   {
-    memset(GlobalType,'\0',MAXCMD);
+    memset(GlobalType,'\0',STRMAX);
   }
 
   strncpy(GlobalParam, S, sizeof(GlobalParam)); // get the parameters, kind of " -A rpm -R fosso -l 1* "
@@ -1026,12 +1047,14 @@ void Usage(char *Name)
   */
 void replace_url_with_auth()
 {
+#define PREFIXMAX 10
+
   const char needle[] = " ";
   const char needle2[] = "//";
   int index = 0;
   char *username = NULL;
   char *password = NULL;
-  char http[10] = "";
+  char http[PREFIXMAX] = "";
   char URI[FILEPATH] = "";
   char *token = NULL;
   char *temp = NULL;
@@ -1039,9 +1062,17 @@ void replace_url_with_auth()
   if (strstr(GlobalParam, "password") && strstr(GlobalParam, "username"))
   {
     temp = strstr(GlobalURL, needle2);
+    if (!temp || (temp - GlobalURL) < 3)
+    {
+      return;
+    }
     strcpy(URI, temp + 2);
-    strncpy(http, GlobalURL, strlen(GlobalURL) - strlen(URI));
+    if (strlen(GlobalURL) - strlen(URI) > PREFIXMAX - 1)
+    {
+      return;
+    }
 
+    strncpy(http, GlobalURL, strlen(GlobalURL) - strlen(URI));
     /* get the first token */
     token = strtok(GlobalParam, needle);
     /* walk through other tokens */
@@ -1052,8 +1083,8 @@ void replace_url_with_auth()
       token = strtok(NULL, needle);
       index++;
     }
-    snprintf(GlobalURL, FILEPATH, "%s%s:%s@%s", http, username, password, URI);
-    memset(GlobalParam,'\0',MAXCMD);
+    snprintf(GlobalURL, URLMAX, "%s%s:%s@%s", http, username, password, URI);
+    memset(GlobalParam,'\0',STRMAX);
   }
 }
 
@@ -1067,11 +1098,11 @@ void MaskPassword()
   int secondIndex = 0;
   char *username = NULL;
   char *token = NULL;
-  char newParam[MAXCMD];
+  char newParam[STRMAX];
   char *beg = NULL;
   char *end = NULL;
 
-  memset(newParam, '\0', MAXCMD);
+  memset(newParam, '\0', STRMAX);
   // SVN if parameters exists
   if (strstr(GlobalParam, "password") && strstr(GlobalParam, "username")) {
     /* get the first token */
@@ -1088,7 +1119,7 @@ void MaskPassword()
     }
     // Create new parameters with masked password
     sprintf(newParam, " --username %s --password ****", username);
-    memset(GlobalParam, '\0', MAXCMD);
+    memset(GlobalParam, '\0', STRMAX);
     strcpy(GlobalParam, newParam);
   }
   // GIT
@@ -1118,30 +1149,60 @@ void MaskPassword()
 char* GetVersionControlCommand(int withPassword)
 {
   char Type[][4] = {"SVN", "Git", "CVS"};
-  char *command = NULL;
-  char TempFileDirectory[MAXCMD];
+  char *command;
+  char *tmpfile_dir;
+  int res;
 
   /** save each upload files in /srv/fossology/repository/localhost/wget/wget.xxx.dir/ */
-  sprintf(TempFileDirectory, "%s.dir", GlobalTempFile);
-
-  command = (char *)malloc(MAXCMD);
-  memset(command,'\0',MAXCMD);
+  res = asprintf(&tmpfile_dir, "%s.dir", GlobalTempFile);
+  if (res == -1)
+  {
+    return NULL;
+  }
 
   if(withPassword < 0) MaskPassword();
   if (0 == strcmp(GlobalType, Type[0]))
   {
     if (GlobalProxy[0] && GlobalProxy[0][0])
-      sprintf(command, "svn --config-option servers:global:http-proxy-host=%s --config-option servers:global:http-proxy-port=%s export %s %s %s --no-auth-cache >/dev/null 2>&1", GlobalProxy[4], GlobalProxy[5], GlobalURL, GlobalParam, TempFileDirectory);
+    {
+      res = asprintf(&command, "svn --config-option servers:global:http-proxy-host=%s --config-option servers:global:http-proxy-port=%s export %s %s %s --no-auth-cache >/dev/null 2>&1", GlobalProxy[4], GlobalProxy[5], GlobalURL, GlobalParam, tmpfile_dir);
+      if (res == -1)
+      {
+        free(tmpfile_dir);
+        return NULL;
+      }
+    }
     else
-      sprintf(command, "svn export %s %s %s --no-auth-cache >/dev/null 2>&1", GlobalURL, GlobalParam, TempFileDirectory);
+    {
+      res = asprintf(&command, "svn export %s %s %s --no-auth-cache >/dev/null 2>&1", GlobalURL, GlobalParam, tmpfile_dir);
+      if (res == - 1)
+      {
+        free(tmpfile_dir);
+        return NULL;
+      }
+    }
   }
   else if (0 == strcmp(GlobalType, Type[1]))
   {
     replace_url_with_auth();
     if (GlobalProxy[0] && GlobalProxy[0][0])
-      sprintf(command, "git config --global http.proxy %s && git clone %s %s %s  && rm -rf %s/.git", GlobalProxy[0], GlobalURL, GlobalParam, TempFileDirectory, TempFileDirectory);
+    {
+      res = asprintf(&command, "git config --global http.proxy %s && git clone %s %s %s  && rm -rf %s/.git", GlobalProxy[0], GlobalURL, GlobalParam, tmpfile_dir, tmpfile_dir);
+      if (res == -1)
+      {
+        free(tmpfile_dir);
+        return NULL;
+      }
+    }
     else
-      sprintf(command, "git clone %s %s %s >/dev/null 2>&1 && rm -rf %s/.git", GlobalURL, GlobalParam, TempFileDirectory, TempFileDirectory);
+    {
+      res = asprintf(&command, "git clone %s %s %s >/dev/null 2>&1 && rm -rf %s/.git", GlobalURL, GlobalParam, tmpfile_dir, tmpfile_dir);
+      if (res == -1)
+      {
+        free(tmpfile_dir);
+        return NULL;
+      }
+    }
   }
 
   return command;

--- a/src/wget_agent/agent/wget_agent.c
+++ b/src/wget_agent/agent/wget_agent.c
@@ -297,6 +297,34 @@ int TaintURL(char *Sin, char *Sout, int SoutSize)
   return(Sin[i]=='\0');
 } /* TaintURL() */
 
+
+/**
+ * \brief Prepare directory for wget.
+ * \param TempFile
+ * \param TempFileDir
+ * \param TempFileDirectory
+ * \parblock
+ * Internal helper function for function GetURL 
+ * \endparblock
+ * \return destination for wget download or NULL
+ */
+char *PrepareWgetDest(char *TempFile, char *TempFileDir, char *TempFileDirectory)
+{
+  if (TempFile && TempFile[0])
+  {
+    /* Delete the temp file if it exists */
+    unlink(TempFile);
+    return TempFileDirectory;
+  }
+  else if(TempFileDir && TempFileDir[0])
+  {
+    return TempFileDir;
+  }
+    
+  return NULL;
+}
+
+
 /**
  * \brief Do the wget.
  * \param TempFile
@@ -376,23 +404,21 @@ int GetURL(char *TempFile, char *URL, char *TempFileDir)
     snprintf(no_proxy, MAXCMD-1, "-e no_proxy='%s'", GlobalProxy[3]);
   }
 
-  if (TempFile && TempFile[0])
-  {
-    /* Delete the temp file if it exists */
-    unlink(TempFile);
+  char *dest;
+
+  dest = PrepareWgetDest(TempFile, TempFileDir, TempFileDirectory);
+
+  if (dest) {
     snprintf(CMD,MAXCMD-1," %s /usr/bin/wget -q %s -P '%s' '%s' %s %s 2>&1",
-        proxy, WgetArgs,TempFileDirectory,TaintedURL,GlobalParam, no_proxy);
-  }
-  else if(TempFileDir && TempFileDir[0])
-  {
-    snprintf(CMD,MAXCMD-1," %s /usr/bin/wget -q %s -P '%s' '%s' %s %s 2>&1",
-        proxy, WgetArgs, TempFileDir, TaintedURL, GlobalParam, no_proxy);
+        proxy, WgetArgs, dest, TaintedURL, GlobalParam, no_proxy);
   }
   else
   {
     snprintf(CMD,MAXCMD-1," %s /usr/bin/wget -q %s '%s' %s %s 2>&1",
-        proxy, WgetArgs,TaintedURL, GlobalParam, no_proxy);
+        proxy, WgetArgs, TaintedURL, GlobalParam, no_proxy);
   }
+
+
   /* the command is like
   ". /usr/local/etc/fossology/Proxy.conf;
      /usr/bin/wget -q --no-check-certificate --progress=dot -rc -np -e robots=off -P

--- a/src/wget_agent/agent/wget_agent.c
+++ b/src/wget_agent/agent/wget_agent.c
@@ -839,60 +839,92 @@ char *PathCheck(char *DirPath)
  */
 int Archivefs(char *Path, char *TempFile, char *TempFileDir, struct stat Status)
 {
-  char CMD[MAXCMD] = {0};
+  char *cmd;
   int rc_system = 0;
+  int res;
 
-  snprintf(CMD,MAXCMD-1, "mkdir -p '%s' >/dev/null 2>&1", TempFileDir);
-  rc_system = system(CMD);
-  if (!WIFEXITED(rc_system))
+  res = asprintf(&cmd , "mkdir -p '%s' >/dev/null 2>&1", TempFileDir);
+  if (res == -1)
   {
-    LOG_FATAL("[%s:%d] Could not create temporary directory", __FILE__, __LINE__);
-    systemError(__LINE__, rc_system, CMD)
+    ASPRINTF_MEM_ERROR_LOG;
     return 0;
   }
 
+  rc_system = system(cmd);
+  if (!WIFEXITED(rc_system))
+  {
+    LOG_FATAL("[%s:%d] Could not create temporary directory", __FILE__, __LINE__);
+    systemError(__LINE__, rc_system, cmd)
+    free(cmd);
+    return 0;
+  }
+  free(cmd);
+
   if (S_ISDIR(Status.st_mode)) /* directory? */
   {
-    memset(CMD,'\0', MAXCMD);
-    snprintf(CMD,MAXCMD-1, "tar %s -cf  '%s' -C '%s' ./ 1>/dev/null", GlobalParam, TempFile, Path);
-    rc_system = system(CMD);
+    res = asprintf(&cmd, "tar %s -cf  '%s' -C '%s' ./ 1>/dev/null", GlobalParam, TempFile, Path);
+    if (res == -1)
+    {
+      ASPRINTF_MEM_ERROR_LOG;
+      return 0;
+    }
+    rc_system = system(cmd);
     if (!WIFEXITED(rc_system))
     {
-      systemError(__LINE__, rc_system, CMD)
+      systemError(__LINE__, rc_system, cmd)
+      free(cmd);
       return 0;
     }
+    free(cmd);
   } else if (strstr(Path, "*"))  // wildcards
   {
-    memset(CMD, '\0', MAXCMD);
     /* for the wildcards upload, keep the path */
     /* copy * files to TempFileDir/temp primarily */
-    snprintf(CMD,MAXCMD-1, "mkdir -p '%s/temp'  > /dev/null 2>&1 && cp -r %s '%s/temp' > /dev/null 2>&1", TempFileDir, Path, TempFileDir);
-    rc_system = system(CMD);
-    if (rc_system != 0)
+    res = asprintf(&cmd, "mkdir -p '%s/temp'  > /dev/null 2>&1 && cp -r %s '%s/temp' > /dev/null 2>&1", TempFileDir, Path, TempFileDir);
+    if (res == -1)
     {
-      systemError(__LINE__, rc_system, CMD)
+      ASPRINTF_MEM_ERROR_LOG;
       return 0;
     }
-    memset(CMD, '\0', MAXCMD);
-    snprintf(CMD,MAXCMD-1, "tar -cf  '%s' -C %s/temp ./  1> /dev/null && rm -rf %s/temp  > /dev/null 2>&1", TempFile, TempFileDir, TempFileDir);
-    rc_system = system(CMD);
+    rc_system = system(cmd);
     if (rc_system != 0)
     {
-      systemError(__LINE__, rc_system, CMD)
+      systemError(__LINE__, rc_system, cmd)
+      free(cmd);
       return 0;
     }
+    free(cmd);
+    res = asprintf(&cmd, "tar -cf  '%s' -C %s/temp ./  1> /dev/null && rm -rf %s/temp  > /dev/null 2>&1", TempFile, TempFileDir, TempFileDir);
+    if (res == -1)
+    {
+      ASPRINTF_MEM_ERROR_LOG;
+      return 0;
+    }
+    rc_system = system(cmd);
+    if (rc_system != 0)
+    {
+      systemError(__LINE__, rc_system, cmd)
+      free(cmd);
+      return 0;
+    }
+    free(cmd);
   } else if(S_ISREG(Status.st_mode)) /* regular file? */
   {
-    memset(CMD, '\0', MAXCMD);
-    snprintf(CMD,MAXCMD-1, "cp '%s' '%s' >/dev/null 2>&1", Path, TempFile);
-    rc_system = system(CMD);
-    if (rc_system != 0)
+    res = asprintf(&cmd, "cp '%s' '%s' >/dev/null 2>&1", Path, TempFile);
+    if (res == -1)
     {
-      systemError(__LINE__, rc_system, CMD)
+      ASPRINTF_MEM_ERROR_LOG;
       return 0;
     }
-  }
-  else return 0; /* neither a directory nor a regular file */
+    rc_system = system(cmd);
+    if (rc_system != 0)
+    {
+      systemError(__LINE__, rc_system, cmd)
+      free(cmd);
+      return 0;
+    }
+    free(cmd);
+  } else return 0; /* neither a directory nor a regular file */
 
   return 1;
 }

--- a/src/wget_agent/agent/wget_agent.c
+++ b/src/wget_agent/agent/wget_agent.c
@@ -482,8 +482,7 @@ int GetURL(char *TempFile, char *URL, char *TempFileDir)
   /* Run from scheduler! store /srv/fossology/repository/localhost/wget/wget.xxx.dir/<files|directories> to one temp file */
   if (TempFile && TempFile[0])
   {
-    char TempFilePath[MAXCMD];
-    memset(TempFilePath,'\0',MAXCMD);
+    char* tmpfile_path;
     /* for one url http://a.org/test.deb, TempFilePath should be /srv/fossology/repository/localhost/wget/wget.xxx.dir/a.org/test.deb */
     int Position = GetPosition(TaintedURL);
     if (0 == Position)
@@ -495,16 +494,23 @@ int GetURL(char *TempFile, char *URL, char *TempFileDir)
       free(delete_tmpdir_cmd);
       SafeExit(26);
     }
-    snprintf(TempFilePath, MAXCMD-1, "%s/%s", TempFileDirectory, TaintedURL + Position);
+    res = asprintf(&tmpfile_path, "%s/%s", TempFileDirectory, TaintedURL + Position);
+    if (res == -1)
+    {
+      ASPRINTF_MEM_ERROR_LOG;
+      free(delete_tmpdir_cmd);
+      SafeExit(ASPRINTF_MEM_ERROR);
+    }
 
-    if (!stat(TempFilePath, &sb))
+    if (!stat(tmpfile_path, &sb))
     {
       if (S_ISDIR(sb.st_mode))
       {
-        res = asprintf(&cmd, "find '%s' -mindepth 1 -type d -empty -exec rmdir {} \\; > /dev/null 2>&1", TempFilePath);
+        res = asprintf(&cmd, "find '%s' -mindepth 1 -type d -empty -exec rmdir {} \\; > /dev/null 2>&1", tmpfile_path);
         if (res == -1)
         {
           ASPRINTF_MEM_ERROR_LOG;
+          free(tmpfile_path);
           free(delete_tmpdir_cmd);
           SafeExit(ASPRINTF_MEM_ERROR);
         }
@@ -512,25 +518,29 @@ int GetURL(char *TempFile, char *URL, char *TempFileDir)
         if (!WIFEXITED(rc_system)) systemError(__LINE__, rc_system, cmd)
         free(cmd);
 
-        res = asprintf(&cmd, "tar -cf  '%s' -C '%s' ./ 1>/dev/null", TempFile, TempFilePath);
+        res = asprintf(&cmd, "tar -cf  '%s' -C '%s' ./ 1>/dev/null", TempFile, tmpfile_path);
         if (res == -1)
         {
           ASPRINTF_MEM_ERROR_LOG;
+          free(tmpfile_path);
           free(delete_tmpdir_cmd);
           SafeExit(ASPRINTF_MEM_ERROR);
         }
       }
       else
       {
-        res = asprintf(&cmd, "mv '%s' '%s' 2>&1", TempFilePath, TempFile);
+        res = asprintf(&cmd, "mv '%s' '%s' 2>&1", tmpfile_path, TempFile);
         if (res == -1)
         {
           ASPRINTF_MEM_ERROR_LOG;
+          free(tmpfile_path);
           free(delete_tmpdir_cmd);
           SafeExit(ASPRINTF_MEM_ERROR);
         }
       }
       
+      free(tmpfile_path);
+
       rc_system = system(cmd);
       if (rc_system != 0)
       {

--- a/src/wget_agent/agent/wget_agent.c
+++ b/src/wget_agent/agent/wget_agent.c
@@ -527,7 +527,6 @@ int GetVersionControl()
 
   int rc = 0;
   int resethome = 0; // 0: default; 1: home is null before setting, should rollback
-  int rc_system =0;
   char *homeenv = NULL;
   int res;
 
@@ -579,8 +578,8 @@ int GetVersionControl()
     */
     LOG_FATAL("please make sure the URL of repo is correct, also add correct proxy for your version control system, command is:%s, GlobalTempFile is:%s, rc is:%d. \n", command, GlobalTempFile, rc);
     /* remove the temp dir /srv/fossology/repository/localhost/wget/wget.xxx.dir/ for this upload */
-    rc_system = system(delete_tmpdir_cmd);
-    if (!WIFEXITED(rc_system)) systemError(__LINE__, rc_system, delete_tmpdir_cmd)
+    rc = system(delete_tmpdir_cmd);
+    if (!WIFEXITED(rc)) systemError(__LINE__, rc, delete_tmpdir_cmd)
     free(tmp_file_directory);
     free(delete_tmpdir_cmd);
     return 1;
@@ -593,16 +592,16 @@ int GetVersionControl()
   {
     systemError(__LINE__, rc, command)
     /* remove the temp dir /srv/fossology/repository/localhost/wget/wget.xxx.dir/ for this upload */
-    rc_system = system(delete_tmpdir_cmd);
-    if (!WIFEXITED(rc_system)) systemError(__LINE__, rc_system, delete_tmpdir_cmd)
+    rc = system(delete_tmpdir_cmd);
+    if (!WIFEXITED(rc)) systemError(__LINE__, rc, delete_tmpdir_cmd)
     LOG_FATAL("DeleteTempDirCmd is:%s\n", delete_tmpdir_cmd);
     free(delete_tmpdir_cmd);
     return 1;
   }
 
   /* remove the temp dir /srv/fossology/repository/localhost/wget/wget.xxx.dir/ for this upload */
-  rc_system = system(delete_tmpdir_cmd);
-  if (!WIFEXITED(rc_system)) systemError(__LINE__, rc_system, delete_tmpdir_cmd)
+  rc = system(delete_tmpdir_cmd);
+  if (!WIFEXITED(rc)) systemError(__LINE__, rc, delete_tmpdir_cmd)
   free(delete_tmpdir_cmd);
 
   return 0; // succeed to retrieve source

--- a/src/wget_agent/agent/wget_agent.c
+++ b/src/wget_agent/agent/wget_agent.c
@@ -25,6 +25,7 @@
 #define _GNU_SOURCE           // for asprintf
 
 #define ASPRINTF_MEM_ERROR  88
+#define ASPRINTF_MEM_ERROR_LOG LOG_FATAL("Not enough memory for asprintf before line %d", __LINE__)
 
 #include "wget_agent.h"
 
@@ -368,7 +369,7 @@ char *PrepareWgetDest(char *TempFile, char *TempFileDir, char *TempFileDirectory
  */
 int GetURL(char *TempFile, char *URL, char *TempFileDir)
 {
-  char CMD[MAXCMD];
+  char *cmd;
   char TaintedURL[MAXCMD];
   char TempFileDirectory[MAXCMD];
   char DeleteTempDirCmd[MAXCMD];
@@ -393,7 +394,6 @@ int GetURL(char *TempFile, char *URL, char *TempFileDir)
     SafeExit(10);
   }
 
-  memset(CMD,'\0',MAXCMD);
   /*
    Wget options:
    --progress=dot :: display a new line as it progresses.
@@ -435,19 +435,25 @@ int GetURL(char *TempFile, char *URL, char *TempFileDir)
   }
 
   char *dest;
+  int res;
 
   dest = PrepareWgetDest(TempFile, TempFileDir, TempFileDirectory);
 
   if (dest) {
-    snprintf(CMD,MAXCMD-1," %s /usr/bin/wget -q %s -P '%s' '%s' %s %s 2>&1",
+    res = asprintf(&cmd," %s /usr/bin/wget -q %s -P '%s' '%s' %s %s 2>&1",
         proxy, WgetArgs, dest, TaintedURL, GlobalParam, no_proxy);
   }
   else
   {
-    snprintf(CMD,MAXCMD-1," %s /usr/bin/wget -q %s '%s' %s %s 2>&1",
+    res = asprintf(&cmd," %s /usr/bin/wget -q %s '%s' %s %s 2>&1",
         proxy, WgetArgs, TaintedURL, GlobalParam, no_proxy);
   }
 
+  if (res == -1)
+  {
+    ASPRINTF_MEM_ERROR_LOG;
+    SafeExit(ASPRINTF_MEM_ERROR);
+  }
 
   /* the command is like
   ". /usr/local/etc/fossology/Proxy.conf;
@@ -455,12 +461,12 @@ int GetURL(char *TempFile, char *URL, char *TempFileDir)
      '/srv/fossology/repository/localhost/wget/wget.xxx.dir/'
      'http://a.org/file' -l 1 -R index.html*  2>&1"
    */
-  LOG_VERBOSE0("CMD: %s", CMD);
-  rc = system(CMD);
+  LOG_VERBOSE0("CMD: %s", cmd);
+  rc = system(cmd);
 
   if (WIFEXITED(rc) && (WEXITSTATUS(rc) != 0))
   {
-    LOG_FATAL("upload %ld Download failed; Return code %d from: %s",GlobalUploadKey,WEXITSTATUS(rc),CMD);
+    LOG_FATAL("upload %ld Download failed; Return code %d from: %s",GlobalUploadKey,WEXITSTATUS(rc),cmd);
     unlink(GlobalTempFile);
     rc_system = system(DeleteTempDirCmd);
     if (!WIFEXITED(rc_system)) systemError(__LINE__, rc_system, DeleteTempDirCmd)
@@ -486,53 +492,80 @@ int GetURL(char *TempFile, char *URL, char *TempFileDir)
 
     if (!stat(TempFilePath, &sb))
     {
-      memset(CMD,'\0',MAXCMD);
       if (S_ISDIR(sb.st_mode))
       {
-        snprintf(CMD,MAXCMD-1, "find '%s' -mindepth 1 -type d -empty -exec rmdir {} \\; > /dev/null 2>&1", TempFilePath);
-        rc_system = system(CMD); // delete all empty directories downloaded
-        if (!WIFEXITED(rc_system)) systemError(__LINE__, rc_system, CMD)
-        memset(CMD,'\0',MAXCMD);
-        snprintf(CMD,MAXCMD-1, "tar -cf  '%s' -C '%s' ./ 1>/dev/null", TempFile, TempFilePath);
+        res = asprintf(&cmd, "find '%s' -mindepth 1 -type d -empty -exec rmdir {} \\; > /dev/null 2>&1", TempFilePath);
+        if (res == -1)
+        {
+          ASPRINTF_MEM_ERROR_LOG;
+          SafeExit(ASPRINTF_MEM_ERROR);
+        }
+        rc_system = system(cmd); // delete all empty directories downloaded
+        if (!WIFEXITED(rc_system)) systemError(__LINE__, rc_system, cmd)
+        free(cmd);
+
+        res = asprintf(&cmd, "tar -cf  '%s' -C '%s' ./ 1>/dev/null", TempFile, TempFilePath);
+        if (res == -1)
+        {
+          ASPRINTF_MEM_ERROR_LOG;
+          SafeExit(ASPRINTF_MEM_ERROR);
+        }
       }
       else
       {
-        snprintf(CMD,MAXCMD-1, "mv '%s' '%s' 2>&1", TempFilePath, TempFile);
+        res = asprintf(&cmd, "mv '%s' '%s' 2>&1", TempFilePath, TempFile);
+        if (res == -1)
+        {
+          ASPRINTF_MEM_ERROR_LOG;
+          SafeExit(ASPRINTF_MEM_ERROR);
+        }
       }
-      rc_system = system(CMD);
+      
+      rc_system = system(cmd);
       if (rc_system != 0)
       {
-        systemError(__LINE__, rc_system, CMD)
+        systemError(__LINE__, rc_system, cmd)
+        free(cmd);
         unlink(GlobalTempFile);
         rc_system = system(DeleteTempDirCmd);
         if (!WIFEXITED(rc_system)) systemError(__LINE__, rc_system, DeleteTempDirCmd)
         SafeExit(24); // failed to store the temperary directory(one file) as one temperary file
       }
+
     }
     else
     {
-      memset(CMD,'\0',MAXCMD);
-      snprintf(CMD,MAXCMD-1, "find '%s' -type f -exec mv {} %s \\; > /dev/null 2>&1", TempFileDirectory, TempFile);
-      rc_system = system(CMD);
+      res = asprintf(&cmd, "find '%s' -type f -exec mv {} %s \\; > /dev/null 2>&1", TempFileDirectory, TempFile);
+      if (res == -1)
+      {
+        ASPRINTF_MEM_ERROR_LOG;
+        SafeExit(ASPRINTF_MEM_ERROR);
+      }
+      rc_system = system(cmd);
       if (rc_system != 0)
       {
-        systemError(__LINE__, rc_system, CMD)
+        systemError(__LINE__, rc_system, cmd)
+        free(cmd);
         unlink(GlobalTempFile);
         rc_system = system(DeleteTempDirCmd);
         if (!WIFEXITED(rc_system)) systemError(__LINE__, rc_system, DeleteTempDirCmd)
         SafeExit(24); // failed to store the temperary directory(one file) as one temperary file
       }
+
     }
   }
 
   if (TempFile && TempFile[0] && !IsFile(TempFile,1))
   {
-    LOG_FATAL("upload %ld File %s not created from URL: %s, CMD: %s",GlobalUploadKey,TempFile,URL, CMD);
+    LOG_FATAL("upload %ld File %s not created from URL: %s, CMD: %s",GlobalUploadKey,TempFile,URL, cmd);
+    free(cmd);
     unlink(GlobalTempFile);
     rc_system = system(DeleteTempDirCmd);
     if (!WIFEXITED(rc_system)) systemError(__LINE__, rc_system, DeleteTempDirCmd)
     SafeExit(15);
   }
+  
+  free(cmd);
 
   /* remove the temp dir /srv/fossology/repository/localhost/wget/wget.xxx.dir/ for this upload */
   rc_system = system(DeleteTempDirCmd);

--- a/src/wget_agent/agent/wget_agent.c
+++ b/src/wget_agent/agent/wget_agent.c
@@ -97,11 +97,13 @@ void DBLoadGold()
   char SQL[MAXCMD];
   long PfileKey;
   char *Path;
-  char SHA256[65], command[PATH_MAX + 13];
+  char SHA256[65];
+  char *cmd;
   FILE *Fin;
   int rc = -1, i;
   PGresult *result;
   int read = 0;
+  int res;
 
   memset(SHA256, '\0', sizeof(SHA256));
 
@@ -109,8 +111,13 @@ void DBLoadGold()
   Fin = fopen(GlobalTempFile,"rb");
 
   // Calculate sha256 value
-  snprintf(command, PATH_MAX + 13, "sha256sum '%s'", GlobalTempFile);
-  FILE* file = popen(command, "r");
+  res = asprintf(&cmd, "sha256sum '%s'", GlobalTempFile);
+  if (res == -1)
+  {
+    SafeExit(ASPRINTF_MEM_ERROR);
+  }
+
+  FILE* file = popen(cmd, "r");
 
   if (!Fin)
   {
@@ -118,23 +125,34 @@ void DBLoadGold()
         GlobalUploadKey,GlobalTempFile,GlobalURL);
     SafeExit(1);
   }
-  if (file != (FILE*) NULL)
+
+  if (!file)
   {
-    read = fscanf(file, "%64s", SHA256);
-    rc = WEXITSTATUS(pclose(file));
+    LOG_FATAL("Unable to get SHA256 from command %s\n", cmd);
+    free(cmd);
+    SafeExit(56);
   }
-  if (file == (FILE*) NULL || rc != 0 || read != 1)
+
+  free(cmd);
+
+  read = fscanf(file, "%64s", SHA256);
+  rc = WEXITSTATUS(pclose(file));
+
+  if (rc || read != 1)
   {
     LOG_FATAL("Unable to calculate SHA256 of %s\n", GlobalTempFile);
     SafeExit(56);
   }
+
   // Change SHA256 to upper case like other checksums
   for (i = 0; i < 65; i++)
   {
     SHA256[i] = toupper(SHA256[i]);
   }
+
   Sum = SumComputeFile(Fin);
   fclose(Fin);
+
   if ((int)ForceGroup > 0)
   {
     rc = chown(GlobalTempFile,-1,ForceGroup);
@@ -147,12 +165,14 @@ void DBLoadGold()
         GlobalUploadKey,GlobalTempFile,GlobalURL);
     SafeExit(2);
   }
+
   if (Sum->DataLen <= 0)
   {
     LOG_FATAL("upload %ld No bytes downloaded from %s to %s.",
         GlobalUploadKey,GlobalURL,GlobalTempFile);
     SafeExit(3);
   }
+
   Unique = SumToString(Sum);
   LOG_VERBOSE0("Unique %s",Unique);
 
@@ -178,6 +198,7 @@ void DBLoadGold()
   {
     Path = GlobalTempFile;
   } /* else if !GlobalImportGold */
+
   LOG_VERBOSE0("Path is %s",Path);
 
   if (!Path)
@@ -186,18 +207,22 @@ void DBLoadGold()
         GlobalUploadKey,Unique);
     SafeExit(5);
   }
+
   LOG_VERBOSE0("Import files %s",Path);
+
   if (fo_RepImport(Path,"files",Unique,1) != 0)
   {
     LOG_FATAL("upload %ld Failed to import %s from %s into files",
         GlobalUploadKey,Unique,Path);
     SafeExit(6);
   }
+
   if ((int)ForceGroup >= 0)
   {
     rc = chown(Path,-1,ForceGroup);
     if (rc) LOG_ERROR("chown failed on %s, error: %s", Path, strerror(errno));
   }
+
   if (Path != GlobalTempFile)
   {
     if(Path)
@@ -233,6 +258,7 @@ void DBLoadGold()
     result = PQexec(pgConn, "SELECT currval('pfile_pfile_pk_seq')");
     if (fo_checkPQresult(pgConn, result, SQL, __FILE__, __LINE__)) SafeExit(182);
   }
+
   PfileKey = atol(PQgetvalue(result,0,0));
   LOG_VERBOSE0("pfile_pk = %ld",PfileKey);
 
@@ -265,6 +291,7 @@ void DBLoadGold()
     free(Sum);
     Sum = NULL;
   }
+
   if (Unique)
   {
     free(Unique);

--- a/src/wget_agent/agent/wget_agent.c
+++ b/src/wget_agent/agent/wget_agent.c
@@ -22,6 +22,9 @@
  * \file wget_agent.c
  * \brief wget_agent: Retrieve a file and put it in the database.
  */
+#define _GNU_SOURCE           // for asprintf
+
+#define ASPRINTF_MEM_ERROR  88
 
 #include "wget_agent.h"
 
@@ -518,13 +521,15 @@ int GetURL(char *TempFile, char *URL, char *TempFileDir)
 int GetVersionControl()
 {
   char *command = NULL;
-  char TempFileDirectory[MAXCMD];
-  char DeleteTempDirCmd[MAXCMD];
-  char TempHome[MAXCMD];
+  char *tmp_file_directory;
+  char *delete_tmpdir_cmd;
+  char *tmp_home;
+
   int rc = 0;
   int resethome = 0; // 0: default; 1: home is null before setting, should rollback
   int rc_system =0;
   char *homeenv = NULL;
+  int res;
 
   homeenv = getenv("HOME");
   if(NULL == homeenv) resethome = 1;
@@ -532,12 +537,28 @@ int GetVersionControl()
   /* We need HOME to point to where .gitconfig is installed
    * path is the repository path and .gitconfig is installed in its parent directory
    */
-  snprintf(TempHome, sizeof(TempHome), "%s/..", fo_config_get(sysconfig, "FOSSOLOGY", "path", NULL));
-  setenv("HOME", TempHome, 1);
+  res = asprintf(&tmp_home, "%s/..", fo_config_get(sysconfig, "FOSSOLOGY", "path", NULL));
+  if (res == -1)
+  {
+    return ASPRINTF_MEM_ERROR;
+  }
+
+  setenv("HOME", tmp_home, 1);
+  free(tmp_home);
 
   /* save each upload files in /srv/fossology/repository/localhost/wget/wget.xxx.dir/ */
-  sprintf(TempFileDirectory, "%s.dir", GlobalTempFile);
-  sprintf(DeleteTempDirCmd, "rm -rf %s", TempFileDirectory);
+  res = asprintf(&tmp_file_directory, "%s.dir", GlobalTempFile);
+  if (res == -1)
+  {
+    return ASPRINTF_MEM_ERROR;
+  }
+
+  res = asprintf(&delete_tmpdir_cmd, "rm -rf %s", tmp_file_directory);
+  if (res == -1)
+  {
+    free(tmp_file_directory);
+    return ASPRINTF_MEM_ERROR;
+  }
 
   command = GetVersionControlCommand(1);
 
@@ -558,26 +579,31 @@ int GetVersionControl()
     */
     LOG_FATAL("please make sure the URL of repo is correct, also add correct proxy for your version control system, command is:%s, GlobalTempFile is:%s, rc is:%d. \n", command, GlobalTempFile, rc);
     /* remove the temp dir /srv/fossology/repository/localhost/wget/wget.xxx.dir/ for this upload */
-    rc_system = system(DeleteTempDirCmd);
-    if (!WIFEXITED(rc_system)) systemError(__LINE__, rc_system, DeleteTempDirCmd)
+    rc_system = system(delete_tmpdir_cmd);
+    if (!WIFEXITED(rc_system)) systemError(__LINE__, rc_system, delete_tmpdir_cmd)
+    free(tmp_file_directory);
+    free(delete_tmpdir_cmd);
     return 1;
   }
 
-  snprintf(command,MAXCMD-1, "tar -cf  '%s' -C '%s' ./ 1>/dev/null", GlobalTempFile, TempFileDirectory);
+  snprintf(command,MAXCMD-1, "tar -cf  '%s' -C '%s' ./ 1>/dev/null", GlobalTempFile, tmp_file_directory);
+  free(tmp_file_directory);
   rc = system(command);
   if (rc != 0)
   {
-    systemError(__LINE__, rc_system, DeleteTempDirCmd)
+    systemError(__LINE__, rc, command)
     /* remove the temp dir /srv/fossology/repository/localhost/wget/wget.xxx.dir/ for this upload */
-    rc_system = system(DeleteTempDirCmd);
-    if (!WIFEXITED(rc_system)) systemError(__LINE__, rc_system, DeleteTempDirCmd)
-    LOG_FATAL("DeleteTempDirCmd is:%s\n", DeleteTempDirCmd);
+    rc_system = system(delete_tmpdir_cmd);
+    if (!WIFEXITED(rc_system)) systemError(__LINE__, rc_system, delete_tmpdir_cmd)
+    LOG_FATAL("DeleteTempDirCmd is:%s\n", delete_tmpdir_cmd);
+    free(delete_tmpdir_cmd);
     return 1;
   }
 
   /* remove the temp dir /srv/fossology/repository/localhost/wget/wget.xxx.dir/ for this upload */
-  rc_system = system(DeleteTempDirCmd);
-  if (!WIFEXITED(rc_system)) systemError(__LINE__, rc_system, DeleteTempDirCmd)
+  rc_system = system(delete_tmpdir_cmd);
+  if (!WIFEXITED(rc_system)) systemError(__LINE__, rc_system, delete_tmpdir_cmd)
+  free(delete_tmpdir_cmd);
 
   return 0; // succeed to retrieve source
 }

--- a/src/wget_agent/agent/wget_agent.h
+++ b/src/wget_agent/agent/wget_agent.h
@@ -47,19 +47,20 @@ typedef struct stat stat_t;
 
 #include "../../ununpack/agent/checksum.h"
 
-#define MAXCMD  2048
+#define URLMAX   3072
+#define STRMAX   2048
 #define FILEPATH 2048
 
-extern char SQL[MAXCMD];
+extern char SQL[STRMAX];
 
 /* for the DB */
 extern PGconn *pgConn;
 /* input for this system */
 extern long GlobalUploadKey;
-extern char GlobalTempFile[MAXCMD];
-extern char GlobalURL[MAXCMD];
-extern char GlobalParam[MAXCMD];
-extern char GlobalType[MAXCMD];
+extern char GlobalTempFile[STRMAX];
+extern char GlobalURL[URLMAX];
+extern char GlobalParam[STRMAX];
+extern char GlobalType[STRMAX];
 extern int GlobalImportGold; /* set to 0 to not store file in gold repository */
 extern gid_t ForceGroup;
 

--- a/src/wget_agent/agent_tests/Unit/wget_agent/testDBLoadGold.c
+++ b/src/wget_agent/agent_tests/Unit/wget_agent/testDBLoadGold.c
@@ -42,9 +42,9 @@ static fo_dbManager* dbManager;
  */
 int  DBLoadGoldInit()
 {
-  char URL[MAXCMD];
-  char TempFileDir[MAXCMD];
-  char TempFile[MAXCMD];
+  char URL[URLMAX];
+  char TempFileDir[STRMAX];
+  char TempFile[STRMAX];
 
   /** -# Create db */
   dbManager = createTestEnvironment(AGENT_DIR, "wget_agent", 1);
@@ -65,8 +65,8 @@ int  DBLoadGoldInit()
   strcpy(GlobalURL, "https://mirrors.kernel.org/fossology/releases/3.0.0/ubuntu/14.04/");
 
   /** -# Delete the record that upload_filename is wget.tar, pre testing */
-  memset(SQL,'\0',MAXCMD);
-  snprintf(SQL,MAXCMD, "DELETE FROM upload where upload_filename = 'fossology.sources.list';");
+  memset(SQL,'\0',STRMAX);
+  snprintf(SQL,STRMAX, "DELETE FROM upload where upload_filename = 'fossology.sources.list';");
   result = PQexec(pgConn, SQL);
   if (fo_checkPQcommand(pgConn, result, SQL, __FILE__ ,__LINE__))
   {
@@ -76,8 +76,8 @@ int  DBLoadGoldInit()
   PQclear(result);
 
   /** -# Insert upload wget.tar */
-  memset(SQL,'\0',MAXCMD);
-  snprintf(SQL,MAXCMD,"INSERT INTO upload (upload_filename,upload_mode,upload_ts) VALUES ('fossology.sources.list',40,now());");
+  memset(SQL,'\0',STRMAX);
+  snprintf(SQL,STRMAX,"INSERT INTO upload (upload_filename,upload_mode,upload_ts) VALUES ('fossology.sources.list',40,now());");
   result = PQexec(pgConn, SQL);
   if (fo_checkPQcommand(pgConn, result, SQL, __FILE__ ,__LINE__))
   {
@@ -86,8 +86,8 @@ int  DBLoadGoldInit()
   }
   PQclear(result);
   /** -# Get upload id */
-  memset(SQL,'\0',MAXCMD);
-  snprintf(SQL,MAXCMD,"SELECT upload_pk from upload where upload_filename = 'fossology.sources.list';");
+  memset(SQL,'\0',STRMAX);
+  snprintf(SQL,STRMAX,"SELECT upload_pk from upload where upload_filename = 'fossology.sources.list';");
   result = PQexec(pgConn, SQL);
   if (fo_checkPQresult(pgConn, result, SQL, __FILE__ ,__LINE__))
   {
@@ -100,8 +100,8 @@ int  DBLoadGoldInit()
   GError* error = NULL;
   char* foConf = get_confFile();
 
-  char cmd[MAXCMD+1];
-  snprintf(cmd, MAXCMD, "sed -i 's|depth.*|depth=3|' %s", foConf);
+  char cmd[STRMAX+1];
+  snprintf(cmd, STRMAX, "sed -i 's|depth.*|depth=3|' %s", foConf);
   if (system(cmd) != 0) {
     printf("cannot reset depth to 3 with %s\n", cmd);
     return 1;
@@ -121,10 +121,10 @@ int  DBLoadGoldInit()
  */
 int DBLoadGoldClean()
 {
-  memset(GlobalTempFile, 0, MAXCMD);
-  memset(GlobalURL, 0, MAXCMD);
-  memset(GlobalParam, 0, MAXCMD);
-  char TempFileDir[MAXCMD];
+  memset(GlobalTempFile, 0, STRMAX);
+  memset(GlobalURL, 0, URLMAX);
+  memset(GlobalParam, 0, STRMAX);
+  char TempFileDir[STRMAX];
 
   strcpy(TempFileDir, "./test_result");
   if (file_dir_existed(TempFileDir))
@@ -136,8 +136,8 @@ int DBLoadGoldClean()
     fo_config_free(sysconfig);
   }
 
-  char repoDir[MAXCMD+1];
-  if (snprintf(repoDir, MAXCMD, "%s/repo", get_sysconfdir())>0) {
+  char repoDir[STRMAX+1];
+  if (snprintf(repoDir, STRMAX, "%s/repo", get_sysconfdir())>0) {
     RemoveDir(repoDir);
   }
 
@@ -175,12 +175,12 @@ void testDBLoadGold()
   //printf("db start\n");
   DBLoadGold();
   //printf("db end\n");
-  char SQL[MAXCMD];
+  char SQL[STRMAX];
   char *pfile_sha1;
   char *pfile_md5;
-  memset(SQL, 0, MAXCMD);
+  memset(SQL, 0, STRMAX);
   PGresult *result;
-  snprintf(SQL, MAXCMD-1, "select pfile_sha1, pfile_md5 from pfile where pfile_pk in (select pfile_fk from "
+  snprintf(SQL, STRMAX-1, "select pfile_sha1, pfile_md5 from pfile where pfile_pk in (select pfile_fk from "
       "upload where upload_pk = %ld);", GlobalUploadKey);
   result =  PQexec(pgConn, SQL); /* SELECT */
   if (fo_checkPQresult(pgConn, result, SQL, __FILE__, __LINE__))
@@ -194,8 +194,8 @@ void testDBLoadGold()
   string_tolower(pfile_sha1);
   string_tolower(pfile_md5);
   //printf("pfile_sha1, pfile_md5 are:%s, %s\n", pfile_sha1, pfile_md5 );
-  char file_name_file[MAXCMD] = {0};
-  char file_name_gold[MAXCMD] = {0};
+  char file_name_file[STRMAX] = {0};
+  char file_name_gold[STRMAX] = {0};
   char string0[3] = {0};
   char string1[3] = {0};
   char string2[3] = {0};

--- a/src/wget_agent/agent_tests/Unit/wget_agent/testSetEnv.c
+++ b/src/wget_agent/agent_tests/Unit/wget_agent/testSetEnv.c
@@ -36,9 +36,9 @@ static char TempFileDir[MAX_LENGTH];
 int  SetEnvInit()
 {
   GlobalUploadKey = -1;
-  memset(GlobalTempFile, 0, MAXCMD);
-  memset(GlobalURL, 0, MAXCMD);
-  memset(GlobalParam, 0, MAXCMD);
+  memset(GlobalTempFile, 0, STRMAX);
+  memset(GlobalURL, 0, URLMAX);
+  memset(GlobalParam, 0, STRMAX);
   return 0;
 }
 
@@ -48,9 +48,9 @@ int  SetEnvInit()
 int  SetEnvClean()
 {
   GlobalUploadKey = -1;
-  memset(GlobalTempFile, 0, MAXCMD);
-  memset(GlobalURL, 0, MAXCMD);
-  memset(GlobalParam, 0, MAXCMD);
+  memset(GlobalTempFile, 0, STRMAX);
+  memset(GlobalURL, 0, URLMAX);
+  memset(GlobalParam, 0, STRMAX);
   return 0;
 }
 /**

